common/debounce/callback.go

```diff
@@ -0,0 +1,122 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package debounce
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/clock"
+)
+
+type DebouncedCallback struct {
+	sync.Mutex
+
+	status          common.DaemonStatusManager
+	lastHandlerCall time.Time
+	callback        func()
+	interval        time.Duration
+	timeSource      clock.TimeSource
+	updateCh        chan struct{}
+
+	loopContext context.Context
+	cancelLoop  context.CancelFunc
+	waitGroup   sync.WaitGroup
+}
+
+// NewDebouncedCallback creates a new DebouncedCallback which will
+// accumulate calls to .Handler function and will only call `callback` on given interval if there were any calls to
+// .Handler in between. It is almost like rate limiter, but it never ignores calls to handler - only postpone.
+func NewDebouncedCallback(timeSource clock.TimeSource, interval time.Duration, callback func()) *DebouncedCallback {
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	return &DebouncedCallback{
+		timeSource:  timeSource,
+		interval:    interval,
+		callback:    callback,
+		updateCh:    make(chan struct{}, 1),
+		loopContext: ctx,
+		cancelLoop:  cancelFn,
+	}
+}
+
+func (d *DebouncedCallback) Start() {
+	if !d.status.TransitionToStart() {
+		return
```

Comment:
tbh I'd treat these as major errors.  mis-calling lifecycle events like this is generally a sign of fatal flaws somewhere

Reply:
what do you mean by "tbh I'd treat these as major errors".
They shouldn't cause errors/panics. In my mind, we should just ignore double-Start, double-Stop

Reply:
why would duplicate starting be a _desirable_ thing to allow though?  by allowing it, you're requiring every future modifier to figure out if that's needed, and that kind of thing tends to be quite a lot of work because it's internal state.

this kind of unused-lifecycle-flexibility tends to become a major complicator in the long run.  if we're not getting some concrete benefit from it, we really shouldn't build it.

Reply:
For clarification here: part of this is just that I really hate that daemon API / the atomics-pattern through the codebase.  
But making a wrapper for it makes plenty of sense, tests look good, no complaint at all about that.

It's more that things using the Daemon API _almost always_ have clear expectations about who should call what, when, but are doing nothing to enforce that.  That kind of thing has repeatedly led to unnecessary calls making it extremely hard to figure out what's _actually_ happening, and I really don't want to make that situation worse :\

_But_ that's not really an issue for an isolated thing used in one place.  And we already have this pattern, and continuing it isn't immediately damaging or anything.
So no request to make changes, just complaining out loud :)